c7n/resolver.py

```diff
@@ -130,8 +153,30 @@ def __init__(self, data, manager):
         }
         self.data = format_string_values(data, **config_args)
         self.manager = manager
+        self.session_factory = manager.session_factory
         self.cache = manager._cache or NullCache({})
-        self.resolver = URIResolver(manager.session_factory, self.cache)
+        self.resolver = URIResolver(self.session_factory, self.cache)
+
+    def _read_value_from_aws_secretsmanager(self, secret: str) -> str:
+        client = self.session_factory().client(
+            'secretsmanager',
+            region_name=self.manager.config.region
+        )
+        secret_string = client.get_secret_value(SecretId=secret)['SecretString']
+        return secret_string
+
+    def _get_headers(self) -> dict:
+        raw_headers_data = self.data.get('headers', {})
+
+        headers = {}
+        for k, v in raw_headers_data.items():
+            if isinstance(v, str):
+                headers[k] = v
+            elif isinstance(v, dict):
+                secret = v.get('value_from')
+                value = self._read_value_from_aws_secretsmanager(secret)
+                headers[k] = value
+        return headers
```

Comment:
Hi Paladin-Dranser, thank you for the PR. I've been wanting this feature for quite a while.
I have created another PR for your branch to address the case when `v` is neither a string nor a dictionary. Please review it at https://github.com/paladin-dranser/cloud-custodian/pull/1.

Reply:
I updated the schema of `value_from` structure, it should not validate a policy if there is incorrect type of header value (see line 128). I'm not sure if there is any reason to do extra action in code implementation in that case. Anyway, I guess it's dangerous to allow to use unsupported types (any reason to allow that?).

Will review your PR in any case.

Reply:
I believe you're referring to the verification below. You're right, at the moment, there is no need to "do an extra action in code implementation." I am thinking that in the future, people might allow more value types (which we don't know what they could be). When that happens, those values could be filtered out if they haven't noticed this PR's `def _get_headers(self)` implementation.
https://github.com/cloud-custodian/cloud-custodian/blob/5d85b3d63850d7f5e2a8b99333ffbc0900dd9684/c7n/resolver.py#L138

Reply:
I told about 'headers' schema, not only `pattern` check:
```
            'headers': {
                'type': 'object',
                'patternProperties': {
                    '': {
                        'oneOf': [
                            {'type': 'string'},
                            {
                                'type': 'object',
                                'required': ['value_from'],
                                'additionalProperties': 'False',
                                'properties': {
                                    'value_from': {
                                        'type': 'string',
                                        'pattern': '^arn:aws:secretsmanager:.+:\\d+:secret:.+$'
                                    },
                                },
                            },
                        ]
                    }
                }
            },
```
From this schema, only `string` and `object` (only `value_from` key is allowed) type are allowed for `headers`.

> people might allow more value types (which we don't know what they could be). When that happens, those values could be filtered out if they haven't noticed this PR's def _get_headers(self) implementation.

In my opinion, it may happen only if adding of new value types is not tested, otherwise author of change will notice how it works in code.

Since we don't know what new types may be (there may not be new types as well), I'd rather not allow them at all than allow them because it simplifies code (we don't need to cover unpredictable cases because we don't allow them) and code do only what it should.

Since this change is local (I guess no any effects on other c7n functionality) and we don't know what we expect in the future, I vote for current code simplicity (I mean, implement only what we want now as simple as possible).

To be honest, I'm not very interested in this functionality anymore (it took too long to get a PR review; so we use other custom approach on my project now; though) so if you still think your changes are important, just let me know when you finish PR (if you plan to do anything else) to my branch and I just merge them.

Reply:
I understand your point. I think we should park it for now and not worry about the PR I created.

Regarding the frustration of waiting too long for a PR review, I share the same sentiment. The review, merge, and release schedule of c7n does not align well with the pace of our project. However, I have come to accept this as a fact because c7n, like many other open-source projects, is maintained by a few individuals who also have their own livelihoods to consider. We will have to find a way to accommodate this, unfortunately. Nonetheless, I hope we can maintain our passion for contributing to open-source projects.

Reply:
fwiw, best way to get attention on prs is to come to the community meeting or speak up in slack (#development channel).

Reply:
> Nonetheless, I hope we can maintain our passion for contributing to open-source projects, from time to time.

Absolutely correct. This is a reason why I don't close my PR or do any other things. Because someone else always may continue your (in this case, mine) work.

I just wanted to say that this PR is not an important to me any more so I'm ready to share ownership with someone who are ready to work on that because he/she needs it.

I will continue to rebase changes from time to time and fix conflicts if any. But I'm always open to discuss any changes to PR but not ready to do them by myself immediately if it's not about fixing comments (e.g., adding more cases that are out of scope of my change 'get headers from secrets').

> best way to get attention on prs is to come to the community meeting or speak up in slack

Good to know for my next PRs.